cranelift/assembler-x64/src/evex.rs

```diff
@@ -0,0 +1,88 @@
+//! Encoding logic for EVEX instructions.
+
+use crate::api::CodeSink;
+
+/// EVEX prefix is always 4 bytes, byte 0 is 0x62
+pub struct EvexPrefix {
+    byte1: u8,
+    byte2: u8,
+    byte3: u8,
+}
+
+/// The EVEX prefix only ever uses the top bit (bit 3--the fourth bit) of any
+/// HW-encoded register.
+#[inline(always)]
+fn invert_top_bit(enc: u8) -> u8 {
+    (!(enc >> 3)) & 1
+}
+
+//         ┌───┬───┬───┬───┬───┬───┬───┬───┐
+// Byte 1: │ R │ X │ B │ R'│ 0 │ 0 │ m │ m │
+//         ├───┼───┼───┼───┼───┼───┼───┼───┤
+// Byte 2: │ W │ v │ v │ v │ v │ 1 │ p │ p │
+//         ├───┼───┼───┼───┼───┼───┼───┼───┤
+// Byte 3: │ z │ L'│ L │ b │ V'│ a │ a │ a │
+//         └───┴───┴───┴───┴───┴───┴───┴───┘
+
+impl EvexPrefix {
+    /// Construct the [`EvexPrefix`] for an instruction.
+    pub fn new(
+        reg: u8,
+        vvvv: u8,
+        (b, x): (Option<u8>, Option<u8>),
+        ll: (u8, u8),
+        pp: u8,
+        mmmmm: u8,
```

Comment:
```suggestion
        mmm: u8,
```

The EVEX field is 3-bit: `EVEX.mmm`. Makes me think we should check that we're using this `u8` correctly elsewhere.